### PR TITLE
Adds Auth0-based authentication to the app.

### PR DESCRIPTION
**Pull request by jessex**
## Description of the change

Auth is being provided by Auth0. I will provide credentials to everyone working actively on the app so that they can continue to run the app locally and access it in deployed environments.

I confess that I don't know how we'd set up the Zeit integration to be able to use this. There is a secret file that needs to be in the base of the repository for Auth to work which is hidden by `.gitignore`, so we'd need to include that in the deployment to Zeit. Those who have more access, let me know what you think about this! I presume we will have to solve this regardless, assuming we are going to host this in production on Zeit/Now.

A number of files added are copied directly from Auth0 docs. Those which are copied directly were _not_ converted to Typescript because 1) the time investment, and 2) the security risk of getting something wrong. We can leave that port for another time, if ever. New files I wrote around the periphery should be in Typescript, however.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,6 +5,10 @@
 # rather than matching a pattern. This helps makes it easier for us to know when
 # it's okay to remove lines from this file.
 
+# Secrets
+auth_config.json
```
Comment (jessex) [inline .gitignore:9]: I'll need to share this file with folks working on the app.

File: package.json [hunk 2]
```diff
@@ -7,6 +7,8 @@
     "postinstall": "touch node_modules/.bin/.keep"
   },
   "dependencies": {
+    "@auth0/auth0-spa-js": "^1.6.5",
+    "@babel/runtime": "^7.9.2",
```
Comment (jessex) [inline package.json:11]: I needed babel/runtime (prod and dev) and babel/polyfill to support the Auth0 SDK.

File: .gitignore [hunk 3]
```diff
@@ -5,6 +5,10 @@
...
+auth_config*.json
```
Comment (szhu) [inline .gitignore:10]: Can you use this instead? See comment above.

```suggestion
/auth_config.json
/auth_config*.json
```

Let me know if that won't work in this case for some reason!

File: src/entry-point-client/App.tsx [hunk 4]
```diff
@@ -1,21 +1,57 @@
...
     <>
       <GlobalStyles />
       <Switch>
-        {PageList.map(({ path, title, contents }) => (
-          <Route key={path} path={path} exact>
-            <WindowTitle>{title}</WindowTitle>
-            {contents}
-          </Route>
-        ))}
+        <Route path="/verify">
```
Comment (jessex) [inline src/entry-point-client/App.tsx:31]: This is listed manually and first instead of being listed in `PageList` to ensure that it will be defined ahead of time in the event that a given route below gets converted into a Redirect to /verify.

File: src/entry-point-client/index.tsx [hunk 5]
```diff
@@ -1,20 +1,51 @@
...
-    <App />
-  </BrowserRouter>
+  <Auth0Provider
+    domain={config.domain}
+    client_id={config.clientId}
+    audience={config.audience}
+    redirect_uri={window.location.origin}
+    onRedirectCallback={onRedirectCallback}
+  >
+    <Router history={history}>
```
Comment (jessex) [inline src/entry-point-client/index.tsx:34]: We needed to switch from `BrowserRouter` to `Router` with the `history` attribute to support proper auth'd navigation back and forth.

File: src/loading-spinner/Loading.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+const override = {
+  display: 'block',
+  margin: '0 auto',
+  position: "fixed",
+  top: "50%",
+  left: "50%",
+  transform: "translate(-50%, -50%)",
+};
+
+const Loading: React.FC = () => {
```
Comment (jessex) [inline src/loading-spinner/Loading.tsx:13]: Here is a general purpose loading spinner that anyone can use in any asynchronous loading phase. I saw one such place in a PR from @macfarlandian earlier. We should standardize on this as the loading spinner. (UI approved by @sychang)

File: src/serviceWorker.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
+// This optional code is used to register a service worker.
```
Comment (jessex) [inline src/serviceWorker.js:1]: Copied from Auth0.

File: src/react-auth0-spa.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
+import createAuth0Client from "@auth0/auth0-spa-js";
```
Comment (jessex) [inline src/react-auth0-spa.js:1]: Copied from Auth0.

File: src/private-route/PrivateRoute.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+import React, { useEffect } from "react";
```
Comment (jessex) [inline src/private-route/PrivateRoute.js:1]: Copied from Auth0.

File: src/verification-needed/VerificationNeeded.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
+import { GlobalStyles } from "../styles";
```
Comment (szhu) [inline src/verification-needed/VerificationNeeded.tsx:1]: If this is a page, can we move the folder to be `src/page-verification-needed`? This makes it so that when viewing a diff or other alphabetical listings of files, similar things will go together.

File: src/utils/history.js [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import { createBrowserHistory } from "history";
```
Comment (jessex) [inline src/utils/history.js:1]: Copied from Auth0.

File: src/verification-needed/VerificationNeeded.tsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+              <h1 className="text-2xl sm:text-3xl font-display text-left">
+                Almost There
+              </h1>
+              <p className="leading-7 text-base sm:text-lg my-6">
+                Look for a verification email in your inbox and click the link in that email.
+                After you verify your email, click the button below and you will be able to reach
+                the home page.
+              </p>
+              <ButtonWrapper className="leading-7 text-base sm:text-lg my-6">
+                <StyledButton label="Get started" onClick={() => refreshAndNavigateAway()}>Get started</StyledButton>
```
Comment (jessex) [inline src/verification-needed/VerificationNeeded.tsx:42]: I had to create this button that users on the verification page click after having verified their email because, for reasons I don't yet understand, simply clicking on the links on the page to go elsewhere does nothing after you've been forcibly redirected here (but they are fine if you navigate here manually). I think it has something to do with the way the routes and redirects are defined in `App.tsx` so if anyone has any ideas then let me know.

File: src/serviceWorker.js (hunk 7 above)
Comment (szhu) [inline src/serviceWorker.js:1]: 1. Is this related to auth0? If not, it would be ideal if this was in a different PR
2. I'm assuming this is copied from create-react-app? 
  - Do their assumptions they're making (if any) also hold true in this codebase?
  - Can you add a comment indicating where this was sourced from

File: src/react-auth0-spa.js (hunk 8 above)
Comment (szhu) [inline src/react-auth0-spa.js:1]: Can we put this under `src/page-verification-needed`? The idea is the keep helper components within the folders where they're needed, and then "graduate" them outwards if/when other things need them as well.

If we were to bring this outwards, maybe put in `src/auth`?

File: src/entry-point-client/index.tsx [hunk 13]
```diff
@@ -1,20 +1,51 @@
+import '@babel/polyfill';
```
Comment (macfarlandian) [inline src/entry-point-client/index.tsx:1]: can we do 
```suggestion
import "core-js/stable";
import "regenerator-runtime/runtime";
```
instead per [recommendation](https://babeljs.io/docs/en/babel-polyfill) or does that break the Auth0 integration?

File: src/private-route/PrivateRoute.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    if (loading || isAuthenticated) {
+      return;
+    }
+    const fn = async () => {
+      await loginWithRedirect({
+        appState: { targetUrl: window.location.pathname }
+      });
+    };
+    fn();
+  }, [loading, isAuthenticated, loginWithRedirect, path]);
```
Comment (szhu) [inline src/private-route/PrivateRoute.js:19]: Would prefer this instead to be a react hook that private pages can use. Then, the routing system doesn't need to treat private pages specially -- keeping making separate concerns separate.

Also, this file can go into the same folder as whatever folder `react-auth0-spa.js` ends up being in!

File: src/loading-spinner/Loading.tsx [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import { GlobalStyles } from "../styles";
```
Comment (szhu) [inline src/loading-spinner/Loading.tsx:1]: Can you move this into `design-system`? Try to move all globally-used visual things there.

**Comment by notfelineit** [comment]
Hey Joshua! For the secrets, we can try adding them to our prod environment using the now cli.

**Comment by jessex** [comment]
@notfelineit if you have some docs or a command you can point me to, I'm happy to give it a shot!

File: src/entry-point-client/index.tsx [hunk 16]
```diff
@@ -1,20 +1,51 @@
...
 if (process.env.NODE_ENV === "development") {
   ReactDOM.render(element, container);
 } else {
   ReactDOM.hydrate(element, container);
 }
+
+// If you want your app to work offline and load faster, you can change
+// unregister() to register() below. Note this comes with some pitfalls.
+// Learn more about service workers: https://bit.ly/CRA-PWA
+serviceWorker.unregister();
```
Comment (szhu) [inline src/entry-point-client/index.tsx:51]: Hm… what's going on here? Did you previously have a CRA app running on this port that didn't get unregistered, and you're trying to do so?

If that's the case, I think more appropriate would be a piece of code to unregister _all_ service workers associated with this origin, rather than including the CRA service worker and unregistering it.

Comment (szhu) [reply src/entry-point-client/index.tsx:51]: Also, consider putting this inside `process.env.NODE_ENV === "development"`. Production shouldn't have this problem.

File: src/serviceWorker.js (hunk 7 above)
Comment (macfarlandian) [reply src/serviceWorker.js:1]: I don't see anything Auth0-specific in here so I'm pretty sure you can just kill it and not bother importing it above ... this is just `create-react-app` boilerplate

File: src/utils/history.js [hunk 17]
```diff
@@ -0,0 +1,3 @@
...
+
+export default createBrowserHistory();
```
Comment (szhu) [inline src/utils/history.js:3]: 1. Can this be a typescript file?
2. Can we rename this folder to not be "utils"? In my experience, the `utils` folder eventually just becomes a big list of items that have about as much organization as its parent folder. Prefer something like `src/history` or `src/entry-point-client`.

**Comment by notfelineit** [comment]
For sure! https://zeit.co/docs/v2/build-step#using-environment-variables-and-secrets <-- you can follow these steps to add the auth0 config as a secret. Can also help after work!

File: src/verification-needed/VerificationNeeded.tsx (hunk 12 above)
Comment (macfarlandian) [reply src/verification-needed/VerificationNeeded.tsx:42]: assuming we wind up still needing this, instead of the click handler can this be a `<Link to="/" />`, as seen in the site header?

(I'm not really familiar with the router package we are using here so someone else may have a more satisfactory solution to the larger issue)

File: src/serviceWorker.js (hunk 7 above)
Comment (szhu) [reply src/serviceWorker.js:1]: I didn't see these comments until after I submitted my review, sorry!

Comment (jessex) [reply src/serviceWorker.js:1]: @macfarlandian as far as I can tell, this doesn't exist in the repo anywhere and needs to be invoked where I've invoked it. Let me know if I'm misunderstanding!

File: src/verification-needed/VerificationNeeded.tsx (hunk 12 above)
Comment (jessex) [reply src/verification-needed/VerificationNeeded.tsx:42]: I tried that to no avail, sadly.

File: src/serviceWorker.js (hunk 7 above)
Comment (macfarlandian) [reply src/serviceWorker.js:1]: it doesn't exist anywhere, but I think all this code is doing is importing this module and immediately disabling it (which is also part of the CRA boilerplate code). Unless taking it out (including killing that import and any references to it) causes the auth to stop working, it shouldn't be needed. 

That said it shouldn't be hurting anything either, it's just cruft, so I'd also be fine with just leaving a TODO somewhere to investigate and clean up so we can move on to more critical stuff

File: src/verification-needed/VerificationNeeded.tsx (hunk 12 above)
Comment (macfarlandian) [reply src/verification-needed/VerificationNeeded.tsx:42]: alas

File: src/loading-spinner/Loading.tsx (hunk 15 above)
Comment (jessex) [reply src/loading-spinner/Loading.tsx:1]: Good call, done!

File: src/utils/history.js (hunk 17 above)
Comment (jessex) [reply src/utils/history.js:3]: Both done!

File: src/verification-needed/VerificationNeeded.tsx (hunk 10 above)
Comment (jessex) [reply src/verification-needed/VerificationNeeded.tsx:1]: Done!

File: src/react-auth0-spa.js (hunk 8 above)
Comment (jessex) [reply src/react-auth0-spa.js:1]: I put all of the Auth0 files in `src/auth` with a brief README.

File: src/private-route/PrivateRoute.js (hunk 14 above)
Comment (jessex) [reply src/private-route/PrivateRoute.js:19]: I filed https://github.com/Recidiviz/covid19-dashboard/issues/27 to do this in the future. Good suggestion.

File: src/entry-point-client/index.tsx (hunk 13 above)
Comment (jessex) [reply src/entry-point-client/index.tsx]: Turns out we only need the regenerator-runtime import, not the core-js import. Thanks!

**Comment by jessex** [comment]
This is just about ready to merge, but I am having trouble getting `yarn build` to work. Any ideas, @szhu?

**Comment by jessex** [comment]
With much late night help from @szhu, this PR is ready for a final review from anyone interested, particularly in some of the build changes that went in in the last few commits from Sean. There is still some minor configuration to set in Auth0 itself to enable everything to work outside of local but that will come soon.

File: package.json [hunk 18]
```diff
@@ -19,6 +21,7 @@
     "@typescript-eslint/eslint-plugin": "2.25.0",
     "@typescript-eslint/parser": "2.25.0",
     "babel-plugin-styled-components": "1.10.7",
+    "babel-polyfill": "^6.26.0",
```
Comment (macfarlandian) [inline package.json:24]: I think you don't need this anymore but we probably should import `regenerator-runtime` explicitly (i also had to do this in my open PR so it'll get dealt with one way or another if you want to just merge)

Comment (jessex) [reply package.json:24]: In the interest of moving forward, I'm going to merge as is but I'm happy to circle back on this if it doesn't come naturally from your open PR!
